### PR TITLE
Failed headrevs become enemies of the state

### PR DESCRIPTION
**Pull request by tralezab**
## About The Pull Request

Headrevs now gain a solo antagonist status if they are exiled instead of killed in a failed revolution. They have two objectives that conflict with eachother: exile (survive off zlevel and don't go to centcom) and hijack (steal the shuttle). They must pick one to be considered victorious.

## Why It's Good For The Game

Right now: Revolution ends, all revs are deantagged, exiled headrevs are now nonantagonists?? Admin headache because this headrev is not an antag but still facing the conseqences of being one

With this change: Revolution ends, exiled headrevs become enemies of the state. they have two options. Live it out off station as an exile, or one last big bang. Show Nanotrasen. Maybe the syndicate won't kill you with such a badass display of power?

## Changelog
:cl:
add: Headrevs become enemies of the state when the revolution is over!
/:cl:

### CODE REVIEW

File: code/game/gamemodes/objective.dm [hunk 1]
```diff
@@ -515,6 +517,20 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
...
 		return TRUE
 
+/datum/objective/exile
+	name = "exile"
+	explanation_text = "Stay alive off station. Do not go to centcom."
+
+/datum/objective/exile/check_completion()
+	var/list/owners = get_owners()
+	for(var/datum/mind/mind as anything in owners)
+		if(!considered_alive(mind)) //dead
```
Comment (Mothblocks) [inline code/game/gamemodes/objective.dm:527]: Useless comment.

File: code/game/gamemodes/objective.dm [hunk 2]
```diff
@@ -515,6 +517,20 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
 			return FALSE
 		return TRUE
 
+/datum/objective/exile
+	name = "exile"
+	explanation_text = "Stay alive off station. Do not go to centcom."
```
Comment (Mothblocks) [inline code/game/gamemodes/objective.dm:522]:
```suggestion
	explanation_text = "Stay alive off station. Do not go to CentCom."
```

File: code/game/gamemodes/objective.dm [hunk 3]
```diff
@@ -515,6 +517,20 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
...
+
+/datum/objective/exile/check_completion()
+	var/list/owners = get_owners()
+	for(var/datum/mind/mind as anything in owners)
+		if(!considered_alive(mind)) //dead
+			return FALSE
+		var/mob/living/alive_current = mind.current
+		if(SSmapping.level_has_any_trait(alive_current.z, list(ZTRAIT_STATION, ZTRAIT_CENTCOM))) //went to centcom or ended round on station
```
Comment (Mothblocks) [inline code/game/gamemodes/objective.dm:530]: This variable is unnecessary, you can just:

```suggestion
		if(SSmapping.level_has_any_trait(mind.current.z, list(ZTRAIT_STATION, ZTRAIT_CENTCOM))) //went to centcom or ended round on station
```

File: code/modules/antagonists/revolution/enemy_of_the_state.dm [hunk 4]
```diff
@@ -0,0 +1,61 @@
+
+/datum/antagonist/enemy_of_the_state
+	name = "enemy of the state"
+	show_in_antagpanel = FALSE
+	show_name_in_check_antagonists = TRUE
+	hijack_speed = 2 //not like they have much to do
+
+/datum/antagonist/enemy_of_the_state/proc/forge_objectives()
+
+	var/datum/objective/exile/choice_1 = new
```
Comment (Mothblocks) [inline code/modules/antagonists/revolution/enemy_of_the_state.dm:10]: `exile_choice`

File: code/modules/antagonists/revolution/enemy_of_the_state.dm [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	choice_1.owner = owner
+	choice_1.objective_name = "Choice"
+	objectives += choice_1
+
+	var/datum/objective/hijack/choice_2 = new
```
Comment (Mothblocks) [inline code/modules/antagonists/revolution/enemy_of_the_state.dm:15]: `hijack_choice`

**Review by Mothblocks — CHANGES_REQUESTED** [review]
As this seems to be built on top of policy, I'm going to pend head admin approval.

**Comment by necromanceranne** [comment]
Be sure to update their hijack speed for them to be able to complete the objective.

**Comment by Mothblocks** [comment]
they did

File: code/modules/antagonists/revolution/enemy_of_the_state.dm [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+	objectives += exile_choice
+
+	var/datum/objective/hijack/hijack_choice = new
+	hijack_choice.owner = owner
+	hijack_choice.objective_name = "Choice"
+	objectives += hijack_choice
+
+/datum/antagonist/enemy_of_the_state/on_gain()
+	//needs to sleep for a tick so the old antag datum can begone before it announces objectives.
+	sleep(1)
```
Comment (Mothblocks) [inline code/modules/antagonists/revolution/enemy_of_the_state.dm:26]: :-1: Don't do this in on_gain, randomly yielding code is terrible. Just do it wherever the actual old antag datum is removed.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/modules/antagonists/revolution/revolution.dm [hunk 7]
```diff
@@ -234,6 +236,8 @@
 
 	if((ishuman(owner.current)))
 		if(owner.current.stat != DEAD)
+			if(deconversion_reason == DECONVERTER_STATION_WIN)
```
Comment (Mothblocks) [inline code/modules/antagonists/revolution/revolution.dm:239]: Move this up to the beginning of `farewell`, it has no reason to be this deep in the chain.

**Comment by Mothblocks** [comment]
Headmins gave approval over Discord.

**Pull request merged**